src/algorithms/piecewise_linear_stretching.jl

```diff
@@ -0,0 +1,351 @@
+    return PiecewiseLinearStretching(tuple(src_knots...), dst_knots; saturate = saturate)
+end
+
+function PiecewiseLinearStretching(src_knots::T, dst_percentile::Percentiles, img::GenericGrayImage; saturate::Bool = true) where T <: Union{Tuple{Vararg{Union{Real,AbstractGray}}}, AbstractVector}
+    if length(src_knots) != length(dst_percentile.p)
+        throw(ArgumentError("The number of src knots and destination percentiles must match."))
+    end
+    dst_knots = percentile(vec(img), collect(dst_percentile.p))
+    return PiecewiseLinearStretching(src_knots, tuple(dst_knots...); saturate = saturate)
+end
+
+function PiecewiseLinearStretching(src_knots::AbstractVector, dst_knots::AbstractVector; saturate::Bool = true)
+    return PiecewiseLinearStretching(tuple(src_knots...), tuple(dst_knots...); saturate = saturate)
+end
+
+function PiecewiseLinearStretching(src_knots::Tuple{Vararg{Union{Real,AbstractGray}}}, dst_knots::Tuple{Vararg{Union{Real,AbstractGray}}}; saturate::Bool = true) 
+    if length(src_knots) != length(dst_knots)
+        throw(ArgumentError("The number of src and destination knots must match."))
+    end
+
+    if length(src_knots) < 2 || length(dst_knots) < 2
+        throw(ArgumentError("You need to specify at least two knots (the start and end of an interval)."))
+    end
+
+    # Promote the src_knots and dst_knots to a common type to address the case
+    # where someone might use a mixture of integers and floating point numbers
+    # when specifying the knots. 
+    T₁ = promote_type(typeof.(src_knots)...)
+    T₂ = promote_type(typeof.(dst_knots)...)
+    T = promote_type(T₁, T₂)
+
+    if saturate 
+        intervals = convert_knots_to_intervals_and_saturate(T.(src_knots), T.(dst_knots))
+    else
+        intervals = convert_knots_to_intervals(T.(src_knots), T.(dst_knots))
+    end   
+
+    return PiecewiseLinearStretching(tuple(intervals...))
+end
+
+function convert_knots_to_intervals(src_knots::NTuple{N, <:Union{Real,AbstractGray}}, dst_knots::NTuple{N, <:Union{Real,AbstractGray}}) where {N}
+    intervals = Vector{Pair{<:AbstractInterval, <: AbstractInterval}}(undef, N-1)
+    for n = 1:(N-1)
+        iₛ = src_knots[n] 
+        jₛ = dst_knots[n] 
+        iₑ = src_knots[n+1]
+        jₑ = dst_knots[n+1]
+        if n == 1
+            interval = ClosedInterval(iₛ, iₑ) => ClosedInterval(jₛ, jₑ)
+            intervals[n] = interval
+        else
+            interval = Interval{:open, :closed}(iₛ, iₑ) => Interval{:open, :closed}(jₛ, jₑ)
+            intervals[n] = interval
+        end
+    end
+    return intervals
+end
+
+function convert_knots_to_intervals_and_saturate(src_knots::NTuple{N, <:Union{Real,AbstractGray}}, dst_knots::NTuple{N, <:Union{Real,AbstractGray}}) where {N}
+    intervals = convert_knots_to_intervals(src_knots, dst_knots)
+    lb = gray(typemin(eltype(src_knots)))
+    ub = gray(typemax(eltype(src_knots)))
+    lb = lb == -Inf ? lb = nextfloat(-Inf) : lb

```

Comment:
```suggestion
    lb = lb == -Inf ? nextfloat(lb) : lb
```
I would have guessed that the extra assignment would have been a bug, but in any case it's best to make this inferrable. Since `-Inf32 == -Inf`, just checking equality does not imply the same type. My implementation will always preserve the type of `lb`.

Likewise for `ub`.